File: sdk-typescript/packages/vault/src/vault.ts
/**
 * WAVE SDK - Vault API
 *
 * Recording storage, VOD management, and archival policies.
 */

import type {
  WaveClient,
  PaginationParams,
  PaginatedResponse,
  Timestamps,
  Metadata,
} from '@wave-av/core';

export type RecordingStatus = "recording" | "processing" | "ready" | "archived" | "failed";
export type StorageTier = "hot" | "warm" | "cold" | "archive";

export interface Recording extends Timestamps {
  id: string;
  organization_id: string;
  stream_id?: string;
  title: string;
  status: RecordingStatus;
  duration_seconds: number;
  file_size_bytes: number;
  format: string;
  resolution?: string;
  frame_rate?: number;
  storage_tier: StorageTier;
  playback_url?: string;
  download_url?: string;
  thumbnail_url?: string;
  tags?: string[];
  metadata?: Metadata;
  expires_at?: string;
}

export interface StorageUsage {
  organization_id: string;
  total_bytes: number;
  hot_bytes: number;
  warm_bytes: number;
  cold_bytes: number;
  archive_bytes: number;
  recording_count: number;
  quota_bytes: number;
  usage_percent: number;
}

export interface ArchivePolicy extends Timestamps {
  id: string;
  name: string;
  tier_after_days: { warm?: number; cold?: number; archive?: number; delete?: number };
  applies_to: "all" | "tagged";
  tags?: string[];
}

export interface UploadSession {
  id: string;
  upload_url: string;
  expires_at: string;
}
export interface TranscodeJob extends Timestamps {
  id: string;
  recording_id: string;
  status: "pending" | "processing" | "ready" | "failed";
  progress_percent: number;
  output_url?: string;
  error?: string;
}

export interface ListRecordingsParams extends PaginationParams {
  status?: RecordingStatus;
  stream_id?: string;
  storage_tier?: StorageTier;
  tags?: string[];
  order_by?: string;
  order?: "asc" | "desc";
}

/**
 * Recording storage, VOD management, transcoding, and archival policies.
 *
 * @example
 * ```typescript
 * const recording = await wave.vault.startRecording(streamId);
 * const usage = await wave.vault.getStorageUsage();
 * const url = await wave.vault.getDownloadUrl(recordingId);
 * ```
 */
export class VaultAPI {
  private readonly client: WaveClient;
  private readonly basePath = "/v1/vault";
  constructor(client: WaveClient) {
    this.client = client;
  }

  async list(params?: ListRecordingsParams): Promise<PaginatedResponse<Recording>> {
    return this.client.get<PaginatedResponse<Recording>>(`${this.basePath}/recordings`, {
      params: params as Record<string, string | number | boolean | undefined>,
    });
  }
  async get(recordingId: string): Promise<Recording> {
    return this.client.get<Recording>(`${this.basePath}/recordings/${recordingId}`);
  }
  async update(
    recordingId: string,
    updates: { title?: string; tags?: string[]; metadata?: Metadata },
  ): Promise<Recording> {
    return this.client.patch<Recording>(`${this.basePath}/recordings/${recordingId}`, updates);
  }
  async remove(recordingId: string): Promise<void> {
    await this.client.delete(`${this.basePath}/recordings/${recordingId}`);
  }
  async getStorageUsage(): Promise<StorageUsage> {
    return this.client.get<StorageUsage>(`${this.basePath}/storage`);
  }
  async createUpload(request: {
    title: string;
    format: string;
    file_size_bytes: number;
    tags?: string[];
  }): Promise<UploadSession> {
    return this.client.post<UploadSession>(`${this.basePath}/uploads`, request);
  }
  async completeUpload(uploadId: string): Promise<Recording> {
    return this.client.post<Recording>(`${this.basePath}/uploads/${uploadId}/complete`);
  }
  async startRecording(
    streamId: string,
    options?: { title?: string; tags?: string[] },
  ): Promise<Recording> {
    return this.client.post<Recording>(`${this.basePath}/recordings`, {
      stream_id: streamId,
      ...options,
    });
  }
  async stopRecording(streamId: string): Promise<Recording> {
    return this.client.post<Recording>(`${this.basePath}/recordings/stop`, { stream_id: streamId });
  }
  async transcode(
    recordingId: string,
    request: { format: string; resolution?: string; bitrate_kbps?: number },
  ): Promise<TranscodeJob> {
    return this.client.post<TranscodeJob>(
      `${this.basePath}/recordings/${recordingId}/transcode`,
      request,
    );
  }
  async getTranscodeJob(jobId: string): Promise<TranscodeJob> {
    return this.client.get<TranscodeJob>(`${this.basePath}/transcode/${jobId}`);
  }
  async createArchivePolicy(
    policy: Omit<ArchivePolicy, "id" | "created_at" | "updated_at">,
  ): Promise<ArchivePolicy> {
    return this.client.post<ArchivePolicy>(`${this.basePath}/policies`, policy);
  }
  async listArchivePolicies(): Promise<ArchivePolicy[]> {
    return this.client.get<ArchivePolicy[]>(`${this.basePath}/policies`);
  }
  async removeArchivePolicy(policyId: string): Promise<void> {
    await this.client.delete(`${this.basePath}/policies/${policyId}`);
  }
  async getDownloadUrl(recordingId: string): Promise<{ url: string; expires_at: string }> {
    return this.client.get<{ url: string; expires_at: string }>(
      `${this.basePath}/recordings/${recordingId}/download`,
    );
  }
}

export function createVaultAPI(client: WaveClient): VaultAPI {
  return new VaultAPI(client);
}

File: sdk-python/wave_sdk/vault.py
"""WAVE SDK - Vault API. Recording storage, VOD management, and archival."""
from __future__ import annotations
from typing import Any
from pydantic import BaseModel
from wave.client import WaveClient

class Recording(BaseModel):
    id: str; organization_id: str; stream_id: str | None = None; title: str; status: str; duration_seconds: float = 0; file_size_bytes: int = 0; format: str | None = None; storage_tier: str = "hot"; playback_url: str | None = None; download_url: str | None = None; tags: list[str] | None = None; created_at: str; updated_at: str

class VaultAPI:
    def __init__(self, client: WaveClient): self._client = client; self._base = "/v1/vault"
    def list(self, **params: Any) -> dict: return self._client.get(f"{self._base}/recordings", params={k: v for k, v in params.items() if v is not None})
    def get(self, recording_id: str) -> Recording: return Recording(**self._client.get(f"{self._base}/recordings/{recording_id}"))
    def update(self, recording_id: str, **kwargs: Any) -> Recording: return Recording(**self._client.patch(f"{self._base}/recordings/{recording_id}", json=kwargs))
    def remove(self, recording_id: str) -> None: self._client.delete(f"{self._base}/recordings/{recording_id}")
    def get_storage_usage(self) -> dict: return self._client.get(f"{self._base}/storage")
    def create_upload(self, title: str, format: str, file_size_bytes: int, **kwargs: Any) -> dict: return self._client.post(f"{self._base}/uploads", json={"title": title, "format": format, "file_size_bytes": file_size_bytes, **kwargs})
    def complete_upload(self, upload_id: str) -> Recording: return Recording(**self._client.post(f"{self._base}/uploads/{upload_id}/complete"))
    def start_recording(self, stream_id: str, **kwargs: Any) -> Recording: return Recording(**self._client.post(f"{self._base}/recordings", json={"stream_id": stream_id, **kwargs}))
    def stop_recording(self, stream_id: str) -> Recording: return Recording(**self._client.post(f"{self._base}/recordings/stop", json={"stream_id": stream_id}))
    def transcode(self, recording_id: str, format: str, **kwargs: Any) -> dict: return self._client.post(f"{self._base}/recordings/{recording_id}/transcode", json={"format": format, **kwargs})
    def get_download_url(self, recording_id: str) -> dict: return self._client.get(f"{self._base}/recordings/{recording_id}/download")
